pintc/src/predicate/optimize/const_folding.rs

```diff
@@ -30,19 +66,131 @@ pub(crate) fn fold_consts(contract: &mut Contract) {
             let simplified_expr_key = contract
                 .exprs
                 .insert(simplified_expr.clone(), simplified_type.clone());
 
             contract.replace_exprs(Some(pred_key), *old_expr_key, simplified_expr_key);
         }
+
+        if !expr_keys_to_replace.is_empty() {
+            did_fold_const = true;
+        }
+        expr_keys_to_replace = FxHashMap::default();
+    }
+
+    did_fold_const
+}
+
+/// In a given contract, simplify any binary ops that consist of boolean or arithmetic (real and int) identities.
+/// Returns whether or not any identities were folded.
+///
+/// Ex:
+/// true || <expr> is true
+/// true && <expr> is <expr>
+/// false || <expr> is <expr>
+/// false && <expr> is <false>
+/// x + 0 is x
+/// x - 0 is x
+/// x * 0 is 0
+/// x * 1 is x
+/// 0 / x is 0
+/// x / 1 is x
+/// x % 1 is x
+pub(crate) fn fold_identities(contract: &mut Contract) -> bool {
+    let mut expr_keys_to_replace: Vec<(ExprKey, ExprKey)> = vec![];
+    let mut did_fold_const: bool = false;
+
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        for expr_key in contract.exprs(pred_key) {
+            if let Expr::BinaryOp { op, lhs, rhs, .. } = expr_key.get(contract) {
+                let lhs_imm = if let Expr::Immediate { value, .. } = lhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                let rhs_imm = if let Expr::Immediate { value, .. } = rhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                let replacement_expr_key = match (op, lhs_imm, rhs_imm) {
+                    (BinaryOp::LogicalAnd, Some(Immediate::Bool(true)), _) => Some(*rhs),
+
+                    (BinaryOp::LogicalAnd, _, Some(Immediate::Bool(true))) => Some(*lhs),
+
+                    (BinaryOp::LogicalAnd, Some(Immediate::Bool(false)), _) => Some(*lhs),
+
+                    (BinaryOp::LogicalAnd, _, Some(Immediate::Bool(false))) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, Some(Immediate::Bool(true)), _) => Some(*lhs),
+
+                    (BinaryOp::LogicalOr, _, Some(Immediate::Bool(true))) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, Some(Immediate::Bool(false)), _) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, _, Some(Immediate::Bool(false))) => Some(*lhs),
+
+                    (BinaryOp::Add, Some(Immediate::Int(0)), _) => Some(*rhs),
+
+                    (BinaryOp::Add, _, Some(Immediate::Int(0))) => Some(*lhs),
+
+                    (BinaryOp::Add, Some(Immediate::Real(0.0)), _) => Some(*rhs),
+
+                    (BinaryOp::Add, _, Some(Immediate::Real(0.0))) => Some(*lhs),
+
+                    (BinaryOp::Sub, _, Some(Immediate::Int(0))) => Some(*lhs),
+
+                    (BinaryOp::Sub, _, Some(Immediate::Real(0.0))) => Some(*lhs),
```

Comment:
We could technically replace `0 - x` or `0.0 - x` with `-x` right? We would have to insert a new `Expr::UnaryOp`